audio/out/ao_sndio.c

```diff
@@ -0,0 +1,320 @@
+        p->hdl = NULL;
+    }
+    free(p->pfd);
+    p->pfd = NULL;
+    p->playing = false;
+}
+
+static int control(struct ao *ao, enum aocontrol cmd, void *arg)
+{
+    struct priv *p = ao->priv;
+    ao_control_vol_t *vol = arg;
+
+    switch (cmd) {
+    case AOCONTROL_GET_VOLUME:
+        if (!p->havevol)
+            return CONTROL_FALSE;
+        vol->left = vol->right = p->vol * 100 / SIO_MAXVOL;
+        break;
+    case AOCONTROL_SET_VOLUME:
+        if (!p->havevol)
+            return CONTROL_FALSE;
+        sio_setvol(p->hdl, vol->left * SIO_MAXVOL / 100);
+        break;
+    default:
+        return CONTROL_UNKNOWN;
+    }
+    return CONTROL_OK;
+}
+
+static void reset(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    if (p->delay) {
+        /* XXX: some times may block here. Possible on no free buf space. */
+        update(ao);
+        if (!sio_stop(p->hdl)) {
+            MP_ERR(ao, "reset: couldn't sio_stop()\n");
+reinit:
+            /* This is work around.
+             * Without this device will never work again. */
+            MP_WARN(ao, "Force reinitialize audio device.\n");
+            uninit(ao);
+            init(ao);
+            return;
+        }
+        if (!sio_start(p->hdl)) {
+            MP_ERR(ao, "reset: sio_start() fail.\n");
+            goto reinit;
+        }
+        p->delay = 0;
+    }
+}
+
+static bool set_pause(struct ao *ao, bool pause)
+{
+
+    if (pause) {
+        reset(ao);
+    }
+
+    return true;
+}
+
+static void start(struct ao *ao)
+{
+    struct priv *p = ao->priv;
+
+    p->playing = true;
+}
+
+static bool audio_write(struct ao *ao, void **data, int samples)
+{
+    struct priv *p = ao->priv;
+    const size_t size = (samples * ao->sstride);
+    size_t rc;
+
+    rc = sio_write(p->hdl, data[0], size);
+    if (rc != size) {
+        MP_WARN(ao, "audio_write: unexpected partial write: required: %zu, written: %zu\n",
+            size, rc);
+        reset(ao);
+        p->playing = false;
+        return false;
+    }
+    p->delay += (rc / ao->sstride);
+
+    return true;
+}
+
+static void get_state(struct ao *ao, struct mp_pcm_state *state)
+{
+    struct priv *p = ao->priv;
+
+    update(ao);
+
+    /* free_samples */
+    state->free_samples = ((p->par.bufsz - p->delay) / ao->sstride);
+    /* queued_samples */
+    state->queued_samples = p->delay;
+    /* delay */
+    state->delay = (p->delay / (double)p->par.rate);
+    /* playing */
+    state->playing = p->playing;
+}
+
+
+const struct ao_driver audio_out_sndio = {
+    .name      = "sndio",
+    .description = "sndio audio output",
+    .init      = init,
+    .uninit    = uninit,
+    .control   = control,
+    .reset     = reset,
+    .set_pause = set_pause,
+    .start     = start,
+    .write     = audio_write,
+    .get_state = get_state,
+    .priv_size = sizeof(struct priv),
+    .priv_defaults = &(const struct priv) {
+        .hdl = NULL,
+        .playing = false,
+    },
```

Comment:
Since it's initializing everything to 0, it could just be `{ 0 }`...

Why is a default struct needed?

Reply:
I am prefer to not rely to external code for critical things.

Reply:
This initialization should be taken care of in `init()`, these aren't configuration parameters. And you have access to the full code (mpv itself).

Reply:
I agree that it doesn't make sense to override the priv defaults, when the default is what you want. If you just drop `.priv_defaults` completely, mpv zeros your `struct priv`.